airflow/providers/google/cloud/hooks/datafusion.py

```diff
@@ -357,12 +418,45 @@ def list_pipelines(
             )
         return json.loads(response.data)
 
+    def _get_pipeline(
+        self,
+        pipeline_name: str,
+        faux_pipeline_id: str,
+        instance_url: str,
+        namespace: str = "default",
+    ) -> Optional[Dict]:
+        url = os.path.join(
+            instance_url,
+            "v3",
+            "namespaces",
+            namespace,
+            "apps",
+            pipeline_name,
+            "workflows",
+            "DataPipelineWorkflow",
+            "runs",
+        )
+        response = self._cdap_request(url=url, method="GET")
+        if response.status != 200:
+            raise AirflowException(
+                f"Retrieving a pipeline failed with code {response.status}"
+            )
+
+        pipelines_list = json.loads(response.data)
+        for pipe in pipelines_list:
```

Comment:
It looks like this `_get_pipeline` gets called every iteration of the polling while loop.
this means on each poll we loop through all the program runs (which may be numerous).

We should only run this loop to "look up program run by faux id" once to get the "real" CDAP job ID.
This real job ID can be used in polling to get this program run directly with a GET request.

Reply:
@jaketf do I correctly understand that state of program can be acceses under:
```
https://xxx-dot-eun1.datafusion.googleusercontent.com/api/v3/namespaces/default/apps/airflow_test/workflows/DataPipelineWorkflow/runs/5d1a277b-17ae-4cab-b3a6-5d23a2f8f672
```
at least that's how it seems to be done here
https://github.com/GoogleCloudPlatform/terraform-provider-cdap/blob/e16a2c5188bf2299482d5c8261ca01f27cd19a44/cdap/resource_streaming_program_run.go#L244

The problem is that I get 404, any ideas?